README
Aca toda la info sobre este projecto


history
 1005  exit
 1006  hostname server.mydomain.com
 1007  sudo hostname server.mydomain.com
 1008  hostname server.mydomain.com
 1009  domainname
 1010  nisdomainname
 1011  domainname ariel
 1012  sudo domainname ariel
 1013  getdomainname
 1014  exit
 1015  ftp
 1016  ls -lt
 1017  cd Documents/gcn4_ChEC/
 1018  ls -lt
 1019  exit
 1020  ls -lt
 1021  top
 1022  ls -lt
 1023  rm Notebook.py
 1024  ls -lt
 1025  exit
 1026  exitx
 1027  exit
 1028  cd Documents/gcn4_ChEC/
 1029  ls -lt
 1030  ipython notebook 
 1031  ssh aerijman@rhino
 1032  exit
 1033  ssh aerijman@rhino
 1034  exit
 1035  ipython notebook
 1036  exit
 1037  ssh aerijman@rhino
 1038  exit
 1039  ipython notebook
 1040  ssh aerijman@rhino
 1041  exit
 1042  ssh aerijman@rhino
 1043  exit
 1044  ipython noteboook
 1045  ipython notebook
 1046  exit
 1047  ssh aerijman@rhino
 1048  ls -lt
 1049  ipython notebook
 1050  exit
 1051  cd /media/ariel/USB\ DISK/ARIEL______LAB/TATA/
 1052  ls-lt
 1053  ls -lt
 1054  python Gene_NGS_Plot.py ~/Documents/gcn4_ChEC/extremes_14 ~/Documents/gcn4_ChEC/extremes_14.wig ARG3
 1055  python Gene_NGS_Plot.py ~/Documents/gcn4_ChEC/extremes_14.wig ~/Documents/gcn4_ChEC/extremes_14.wig ARG3
 1056  python Gene_NGS_Plot.py ~/Documents/gcn4_ChEC/extremes_14.wig ~/Documents/gcn4_ChEC/extremes_14.wig ILV6
 1057  python Gene_NGS_Plot.py lib/SGD.tsv ~/Documents/gcn4_ChEC/extremes_14.wig ~/Documents/gcn4_ChEC/extremes_14.wig ILV6
 1058  python Gene_NGS_Plot.py lib/SGD.tsv ~/Documents/gcn4_ChEC/extremes_14.wig ~/Documents/gcn4_ChEC/extremes_14.wig ARG3
 1059  python Gene_NGS_Plot.py lib/SGD.tsv ~/Documents/gcn4_ChEC/extremes_14.wig ~/Documents/gcn4_ChEC/extremes_14.wig HIS4
 1060  exit
 1061  ssh aerijman@rhino
 1062  exit
 1063  ssh aerijman@rhino
 1064  exit
 1065  ssh aerijman@rhino
 1066  exit
 1067  cd Documents/
 1068  cd gcn4_ChEC/
 1069  ipython notebook
 1070  ssh aerijman@rhino
 1071  exit
 1072  pip install vagrant
 1073  sudo apt-get install vagrant
 1074  sudo apt-get install virtualbox
 1075  lsb_release-a
 1076  lsb_release -a
 1077  wget -O fsnd-virtual-machine.zip "https://d17h27t6h515a5.cloudfront.net/topher/2016/December/58488015_fsnd-virtual-machine/fsnd-virtual-machine.zip"
 1078  ls -lt
 1079  mv fsnd-virtual-machine.zip Downloads/
 1080  cd Downloads/
 1081  cd ..
 1082  cd Downloads/
 1083  ls -lt
 1084  unzip fsnd-virtual-machine.zip 
 1085  ls -lt
 1086  cd FSND-Virtual-Machine/
 1087  ls -lt
 1088  cd vagrant/
 1089  vagrant up
 1090  vagrant ssh
 1091  nm-tool
 1092  nm -tool
 1093  nmcli -t 
 1094  nmcli -t -f IP4.DNS device show eth0
 1095  ls -lt
 1096  which apache
 1097  which apache2
 1098  which httpd
 1099  sudo apt-get update
 1100  sudo apt-get install apache2
 1101  sudo service apache2 restart
 1102  more /etc/apache2/ports.conf 
 1103  apt-cache search php5-
 1104  sudo apt-get install php5 libapache2-mod-php5-mcrypt
 1105  sudo apt-get install php5-sqlite
 1106  vagrant up
 1107  cd Downloads/
 1108  cd FSND-Virtual-Machine/
 1109  ls -lt
 1110  cd vagrant/
 1111  ls -lt
 1112  vagrant up
 1113  vagrant ssh
 1114  ls -lt
 1115  cd Downloads/
 1116  ls -lt
 1117  cd FSND-Virtual-Machine/
 1118  ls -lt
 1119  cd vagrant/
 1120  ls -lt
 1121  vagrant up
 1122  ls -lt
 1123  cd ../vagrant/forum/
 1124  ls -lt
 1125  which ssh
 1126  more forum.py 
 1127  python
 1128  exit
 1129  python
 1130  exit
 1131  ssh aerijman@rhino
 1132  ipython notebook
 1133  enable remote connections
 1134  more Documents/gcn4_ChEC/ACGT_genome.csv 
 1135  ls -lt
 1136  git --version
 1137  sudo add-apt-repository ppa:git-core/ppa -y
 1138  sudo apt-get update
 1139  sudo apt-get install git -y
 1140  git
 1141  cd Documents/gcn4_ChEC/
 1142  git config --global user.name aerijman
 1143  git config --global user.email aerijman@fredhutch.org
 1144  cd ..
 1145  git init gcn4_ChEC
 1146  cd gcn4_ChEC/
 1147  ls -lt
 1148  nano README
 1149  git add README
 1150  git add ACGT_genome.csv
 1151  git commit -m "edit in README"
 1152  git remote add origin https://github.com/aerijman/gcn4_ChEC.git
 1153  git push origin master
 1154  ls -lt
 1155  git add Notebook.ipynb notebook_library.py Untitled.ipynb 
 1156  git commit -m "edit in README"
 1157  git push origin master
 1158  history
 1159  more wig_totals.csv 
 1160  git add *
 1161  git remove fragments_*
 1162  git rm fragments_*
 1163  history
 1164  git push origin master
 1165  git commit -m "agrege TODOgit push origin master!"
 1166  ls -lt
 1167  more README 
 1168  ls -lt
 1169  history
 1170  cd ..
 1171  ls -lt
 1172  cd gcn4_ChEC/
 1173  ls -lt
 1174  mv fragments_* ..
 1175  ls -lt
 1176  git add *
 1177  git commit -m "agrege TODOgit push origin master!"
 1178  git push origin master
 1179  git add *
 1180  git commit -m "agrege TODOgit push origin master!"
 1181  git push origin master
 1182  git push -f
 1183  git push -f origin master
 1184  sudo apt-get install --reinstall ttf-mscorefonts-installer  
 1185  echo "GGTGCTAATTATGGCATTGATTGGGGCGGTAAAACATTCAAAGTC" | tr "[ACTG]" "[TGAC]" | rev
 1186  ls -lt
 1187  grep "CACAATGCTCCATAACTCTTCAGGCCTTAAACGACCATCCATG" Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1188  grep "CACAATGCTCCATAACTCTTCAGGCCTTAAACGACCATCCATG" Documents/scripts/FindPromoters/libs/chromosomes/chr*
 1189  more Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1190  tr -d '\n' Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa > a
 1191  tr -d '\n' Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1192  tr Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa -d '\n'
 1193  tr Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa -d 
 1194  tr -d '\n' Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1195  echo Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1196  cat Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa 
 1197  cat Documents/scripts/FindPromoters/libs/chromosomes/chr1.fa | tr -d '\n'

notebook_library.py
import sys
import pandas as pd
import numpy as np
from sklearn.neighbors import KernelDensity
from ggplot import *


from numpy import NaN, Inf, arange, isscalar, asarray, array


# Dictionary. Will be used to access each chromosome faster
roman_numbers = {'I':'1','II':'2','III':'3','IV':'4','V':'5','VI':'6','VII':'7',\
'VIII':'8','IX':'9','X':'10','XI':'11','XII':'12','XIII':'13','XIV':'14',\
'XV':'15','XVI':'16','XVII':'17','M':'M'}



def open_sgd(promoter_length):
	
	# open sgd file into a dataframe
	sgd = pd.read_csv('./SGD.tsv', delimiter='\t')

	#assign column names and use only the 'ORF' rows
	sgd.columns = ['n0','ORF_or_what','n1','locus','gene','n2','n3','n4'\
			      ,'chromosome','start','stop','W/C','n5','n6','n7','n8']
	sgd = sgd[sgd['ORF_or_what']=='ORF']
    
	# get rid of non-informative columns and 2-micron chromosome
	sgd = sgd.drop(['ORF_or_what','n0','n1','n2','n3','n4','n5','n6','n7','n8'],1).set_index('locus')
	sgd = sgd.ix[sgd.chromosome!='2-micron']

	# -1 to Creek and 1 to Watson
	sgd.loc[sgd['W/C']=='C', 'W/C']=-1
	sgd.loc[sgd['W/C']=='W', 'W/C']=1
    
	# create the column 'start promoter' and replace negative numbers \
	# with 0 (promoters that are <600bp form the start of the chromosome)
	sgd['start_promoter'] = sgd['start']-(sgd['W/C']*promoter_length)
	sgd.loc[sgd['start_promoter']<0,'start_promoter']=0

	# create the column 'start promoter' and replace negative numbers \
	# with 0 (promoters that are <600bp form the start of the chromosome)
	sgd['promoter_median'] = (sgd['start'] + sgd['start_promoter']) / 2
	sgd = sgd.sort_values(['chromosome','promoter_median'])
	sgd.reset_index(inplace=True)
    
	return sgd



def open_wig(wigfile):

# Open the wig file and make a pointer to where each chromosomes start
    wig, w_pointer = [i.strip('\n').split('\t')[:2] for i in open(wigfile)], {}
    wig2, n = [], 0
    for i in wig[1:]:
        if len(i)==1:
            w_pointer[roman_numbers[i[0][22:]]] = n
            print i
            #wig2.append(i)
            #n+=1
        else:
            if int(i[1])>=0:                   ## era >=20...
                wig2.append(i)
                n+=1
        
    # the last pointer to the end of the file, when looking for a gene in chromosome 17 and there is no NEXT chromosome
    #w_pointer['17']=n 

    # wig into dataframe
    wig = pd.DataFrame(wig2)
    wig.columns = ['position','reads']

    return wig, w_pointer



def retrieve_positions(locus, sgd, wig, w_pointer):

	# chromosomes as pointer for the wig file
	this_chromosome = int(sgd[sgd.index==locus].values[0][1])
	this = w_pointer[str(this_chromosome)]

	# assign the key to NOT to be the current chromosome
	key = [1 if this_chromosome>7 else 10][0]

	# Now go and find me the closest chromosome in the wig file (they are not in order!!)
	for i,j in w_pointer.iteritems():
		if j-this < abs(w_pointer[str(key)]-this) and j>this and i != this_chromosome:
			key = i 
	next = w_pointer[str(key)]

	# promoter positions to find in the wig 
	promoter_positions = [int(i) for i in sgd[sgd.index==locus][['start_promoter', 'start']].as_matrix()[0]]
	
	# In case the gene is in the Creek direction
	if sgd[sgd.index==locus]['W/C'].values[0]==-1:
		promoter_positions = [promoter_positions[1],promoter_positions[0]]

	# restricting the wig to the chromosome and the positions 
	Gene_df =  wig[(wig.index > this) & (wig.index < next)]

	# Only if the wig file contains counts for this gene, otherwise return False!
	if Gene_df.position.empty == False:
		Gene_df = Gene_df[(Gene_df.position.astype(int)>promoter_positions[0]) & (Gene_df.position.astype(int)<promoter_positions[1])]
		return Gene_df
	else:
		return pd.DataFrame([])

    

def kernelPDF(x, y, bandwidth=50, kernel_choose='epanechnikov'):

	# Prepare "histogram-like" data
	histo = []
	for i in range(len(y)):
		for j in range(y[i]):
			histo.append(x[i])

	histo = np.array(histo)
    
	# Kernel Density Estimation with Scikit-learn
	kde = KernelDensity(kernel=kernel_choose, bandwidth=bandwidth).fit(histo[:,np.newaxis])
	pdf = np.exp(kde.score_samples(x[:,np.newaxis]))*10000
        
	return pdf



def ggplotIt(x,y,pdf,kernel_choose='epanechnikov'):
    print_df = pd.DataFrame([x,y,pdf]).T
    print_df.columns = ['position','wig score', kernel_choose]

    print ggplot(print_df, aes('position', 'wig score')) + geom_area() \
      + geom_line(aes('position',kernel_choose), size=2, color='red')

      
        
def peakdet(v, delta, x = None):
    """
    Converted from MATLAB script at http://billauer.co.il/peakdet.html
    
    Returns two arrays
    
    function [maxtab, mintab]=peakdet(v, delta, x)
    %PEAKDET Detect peaks in a vector
    %        [MAXTAB, MINTAB] = PEAKDET(V, DELTA) finds the local
    %        maxima and minima ("peaks") in the vector V.
    %        MAXTAB and MINTAB consists of two columns. Column 1
    %        contains indices in V, and column 2 the found values.
    %      
    %        With [MAXTAB, MINTAB] = PEAKDET(V, DELTA, X) the indices
    %        in MAXTAB and MINTAB are replaced with the corresponding
    %        X-values.
    %
    %        A point is considered a maximum peak if it has the maximal
    %        value, and was preceded (to the left) by a value lower by
    %        DELTA.
    
    % Eli Billauer, 3.4.05 (Explicitly not copyrighted).
    % This function is released to the public domain; Any use is allowed.
    
    """
    maxtab = []
    mintab = []
       
    if x is None:
        x = arange(len(v))
    
    v = asarray(v)
    
    if len(v) != len(x):
        sys.exit('Input vectors v and x must have same length')
    
    if not isscalar(delta):
        sys.exit('Input argument delta must be a scalar')
    
    if delta <= 0:
        sys.exit('Input argument delta must be positive')
    
    mn, mx = Inf, -Inf
    mnpos, mxpos = NaN, NaN
    
    lookformax = True
    
    for i in arange(len(v)):
        this = v[i]
        if this > mx:
            mx = this
            mxpos = x[i]
        if this < mn:
            mn = this
            mnpos = x[i]
        
        if lookformax:
            if this < mx-delta:
                maxtab.append((mxpos, mx))
                mn = this
                mnpos = x[i]
                lookformax = False
        else:
            if this > mn+delta:
                mintab.append((mnpos, mn))
                mx = this
                mxpos = x[i]
                lookformax = True

    return array(maxtab), array(mintab)

def getFromGit(link):

	from html2text import html2text
	import requests
	import numpy as np
	import pandas as pd

	f = requests.get(link)

	matriz=[]
	for i in f.text.split('\n'):
	    cols=[]
	    for j in i.split(','):
	        cols.append(j.encode('ascii','replace'))
	    matriz.append(cols)

	df = pd.DataFrame(matriz)

	df.columns =  np.array(df.loc[0])
	df = df.ix[1:,1:]
	df.set_index('id', inplace=True)
	#df.apply(pd.to_numeric)
	print " DO df.apply(pd.to_numeric) or a slice of df... \n"
	
	return df
	
        


countWig.py
import sys
import pandas as pd
import numpy as np

from sklearn.neighbors import KernelDensity

# 0 - Should have a list of all TSSs and their frecuency. Where frec. is > threshold (50%?), use that TSS.
# 1 - Check percentage of genes that share promoter vs number of bp of promoter from ORF (600bp - 1000bp)
# 	  some peaks are beyond 600bp from ATG... I would use 1000bp for promoter region, but I will use an optimal number
#	  after I see the results.
# 2 - Take out the genes that share peaks for the first analysis. Keep them in a separate list that will be informative later.
# 3 - list of genes with number of reads in their promoter region sorted and check if they can be clustered 
#	  into two or three bins.
# 5 - How many peaks has each gene ?
# 6 - For genes with only one peak. Where does it lay from the TSS, and for all genes, where does the median of all values lay?
# 7 - Compare number of reads from ChEC-seq to the data of MalikaSaint_et_al_2014 (mRNA or P. expression of genome upon SM induction).
# 8 - SAGA pr TFIID?
# 9 - Tables in database with NORMALIZED counts for each gene for gcn4 (+ and -SM) and meidator (+ and -SM)


# Dictionary. Will be used to access each chromosome faster
roman_numbers = {'I':'1','II':'2','III':'3','IV':'4','V':'5','VI':'6','VII':'7',\
'VIII':'8','IX':'9','X':'10','XI':'11','XII':'12','XIII':'13','XIV':'14',\
'XV':'15','XVI':'16','XVII':'17','M':'M'}


sgd_file = './SGD.tsv'
wigfile = sys.argv[1]

promoter_length = 600




from scipy import stats



def main():

	# open sgd file into a DataFrame  --> (index='locus'), columns = ['gene','chromosome','start','stop','W/C','start_promoter'] 
	global sgd
	sgd = open_sgd()

	# open wig file into a dataframe  -->  (index = index -used by w_pointer), columns = ['position','reads']
	global wig, w_pointer 
	wig, w_pointer = open_wig()
	
	counts=[]	

	# Over all genes in SGD
	for i in sgd.itertuples():

		# Check that the gene name is correct and the wig_df is not empty( so there are reads for this gene)
		if type(i.gene)==str:
			Gene_df = retrieve_positions(i.gene)
			if not Gene_df.empty:
				#print i.Index,",",i.gene,",",np.sum(Gene_df.reads.astype(int))
				#counts.append([i.Index,np.sum(Gene_df.reads.astype(int))])
				if i.gene == 'ARG3':
					print kernelPDF(Gene_df.astype(int).as_matrix())	
	#print counts


def open_sgd():
	
	# open sgd file into a dataframe
	sgd = pd.read_csv(sgd_file, delimiter='\t')

	#assign column names and use only the 'ORF' rows
	sgd.columns = ['n0','ORF_or_what','n1','locus','gene','n2','n3','n4'\
			      ,'chromosome','start','stop','W/C','n5','n6','n7','n8']
	sgd = sgd[sgd['ORF_or_what']=='ORF']

	# get rid of non-informative columns and 2-micron chromosome
	sgd = sgd.drop(['ORF_or_what','n0','n1','n2','n3','n4','n5','n6','n7','n8'],1).set_index('locus')
	sgd = sgd.ix[sgd.chromosome!='2-micron']

	# -1 to Creek and 1 to Watson
	sgd.loc[:,'W/C'][sgd['W/C']=='C']=-1
	sgd.loc[:,'W/C'][sgd['W/C']=='W']=1

	# create the column 'start promoter' and replace negative numbers \
	# with 0 (promoters that are <600bp form the start of the chromosome)
	sgd['start_promoter'] = sgd['start']-(sgd['W/C']*promoter_length)
	sgd.loc[:,'start_promoter'][sgd['start_promoter']<0]=0

	return sgd				 


def open_wig():

	# Open the wig file and make a pointer to where each chromosomes start
	wig, w_pointer = [i.strip('\n').split('\t')[:2] for i in open(wigfile)], {}
	for n in range(1,len(wig)):
		if len(wig[n])==1:
			w_pointer[roman_numbers[wig[n][0][22:]]] = n
	
	# the last pointer to the end of the file, when looking for a gene in chromosome 17 and there is no NEXT chromosome
	w_pointer['17']=n 

	# wig into dataframe
	wig = pd.DataFrame(wig)
	wig.columns = ['position','reads']

	return wig, w_pointer
	

def retrieve_positions(gene):
	
	# chromosomes as pointer for the wig file
	this_chromosome = int(sgd[sgd['gene']==gene].values[0][1])
	this = w_pointer[str(this_chromosome)]

	# assign the key to NOT to be the current chromosome
	key = [1 if this_chromosome>7 else 10][0]

	# Now go and find me the closest chromosome in the wig file (they are not in order!!)
	for i,j in w_pointer.iteritems():
		if j-this < abs(w_pointer[str(key)]-this) and j>this and i != this_chromosome:
			key = i 
	next = w_pointer[str(key)]

	# promoter positions to find in the wig 
	promoter_positions = [int(i) for i in sgd[sgd['gene']==gene][['start_promoter', 'start']].as_matrix()[0]]
	
	# In case the gene is in the Creek direction
	if sgd[sgd['gene']==gene]['W/C'].values[0]==-1:
		promoter_positions = [promoter_positions[1],promoter_positions[0]]

	# restricting the wig to the chromosome and the positions 
	Gene_df =  wig[(wig.index > this) & (wig.index < next)]

	# Only if the wig file contains counts for this gene, otherwise return False!
	if Gene_df.position.empty == False:
		Gene_df = Gene_df[(Gene_df.position.astype(int)>promoter_positions[0]) & (Gene_df.position.astype(int)<promoter_positions[1])]
		return Gene_df
	else:
		return pd.DataFrame([])


def kernelPDF(Gene_df, bandwidth=50):

	# extract x and y form the matrix Gene_df
	x,y = Gene_df[:,0], Gene_df[:,1]

	# Prepare "histogram-like" data
	histo = []
	for i in range(len(y)):
		for j in range(y[i]):
			histo.append(x[i])

	histo = np.array(histo)

	# Kernel Density Estimation with Scikit-learn
	kde = KernelDensity(kernel='epanechnikov', bandwidth=bandwidth).fit(histo[:,np.newaxis])
	pdf = np.exp(kde.score_samples(x[:,np.newaxis]))*10000

	return pdf

def kernel_SciPy(Gene_df):
	
	x,y = np.array(Gene_df['position']).astype(int), np.array(Gene_df['reads']).astype(int)

	# Prepare "histogram-like" data
	histo = []
	for i in range(len(y)):
		for j in range(y[i]):
			histo.append(x[i])

	histo = np.array(histo)

	kde = stats.gaussian_kde(x, bw_method = 0.3/y.std(ddof=1))
	
	return kde.evaluate(x)

if __name__ == '__main__':
	main()

